lib/mastodon/plugin.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+module Mastodon
+  class Plugin
+    NamedPath = Struct.new(:name, :path, :type)
+
+    attr_reader :actions, :paths, :assets
+
+    def initialize
+      @actions = Set.new
+      @paths   = Set.new
+      @assets  = Set.new
+    end
+
+    # defined by the plugin
+    def setup!
+      raise NotImplementedError.new, "Please define a setup! method"
+    end
+
+    private
+
+    def use_asset(path, outlet: nil)
+      path = path_prefix(path, relative: true)
+      @assets.add path
+      @paths.add NamedPath.new(outlet, path, :outlet) if outlet
+    end
+
+    def use_asset_directory(glob)
+      use_directory(glob) { |path| use_asset(path) }
+    end
+
+    def use_class(path)
+      @actions.add(proc { require path_prefix path })
+    end
+
+    def use_class_directory(glob)
+      use_directory(glob) { |path| use_class(path) }
+    end
+
+    # Add translation overrides
+    # example usage: `use_translations("locales")`
+    # ^^ search for all .js files (for example en.js) for translation overrides in the
+    # NB: The translation files take the following format:
+    # module.exports = {
+    #   "<translation_key>": "overriding value"
+    # }
+    def use_translations(path)
+      use_directory(path) do |p|
+        @paths.add NamedPath.new(
+          p.split('/').last.gsub('.js', ''),
+          path_prefix(p, relative: true),
+          :locale
+        )
+      end
+    end
+
+    # Add an api route
+    # example usage: `use_route(:get, "examples/:id", "examples#show")`
+    # ^^ create an endpoint at /api/v1/examples/:id which routes to the show action on ExamplesController
+    # NB: be sure to define a controller action by adding a controller or using the 'extend_class' option!
+    def use_route(verb, route, action)
```

Comment:
What if it was something like a `in_router` method that takes a block and executes it in the router context, thereby allowing the plugin to use the Rails routes API fully?